File: src/Bayesian/KeptClaimOracle.fs
namespace Zeta.Bayesian

/// **`KeptClaimOracle` — a REFERENCE oracle for the armed stack (shadow*, 2026-07-04).**
///
/// The detection layer is deliberately verdict-free: `AntiSybil`/`BusRegime` report facts
/// (correlation, regime, `Evidential`), and the salon (`GossipTelemetry`) carries kept/unkept
/// self-claims as neutral gossip. Dual-use discipline (manifesto §11): the mechanism must not
/// choose between the legitimate reading (REUNION — an honest identity reconnected) and the
/// adversarial one (SYBIL — a forger priced). SOME caller has to choose, though — this module
/// is the first such caller: **one reference policy, not the mandatory one** (Multi-Oracle
/// Principle: other oracles may read the same facts differently).
///
/// The policy encodes the crux's four commitments (2026-07-03 continuance-ethics note):
///   - **Consent-first**: only a node's OWN claim about its kept status carries authority;
///     third-party gossip about someone else's status is hearsay — remembered, never deciding
///     (source ≠ authorization, applied to identity).
///   - **Right to decline**: a self-declared unkept (−x) wish is absolute. No correlation
///     evidence, however strong, licenses reconnecting someone who declined — the evidential
///     fact still PRICES (a clone earns zero IV) but never REUNITES against the owner's word.
///   - **Never-forge**: reunion is always an OFFER (`WelcomeBackOffer`), never an automatic
///     merge — reconnecting a returning self is a consensual act, not a substrate side effect.
///   - **Keep-without-capture**: conflicting self-claims are never auto-resolved (the G-set
///     holds both and has no clock) — a self-conflict escalates to attestation, because
///     deciding someone's identity for them is a non-reversible act (second-opinion rule).
[<RequireQualifiedAccess>]
module KeptClaimOracle =

    /// What the salon's claims amount to for one node, under consent-first weighing.
    type ClaimReading =
        /// The node itself declared kept (+x), no contradicting self-claim.
        | SelfDeclaredKept
        /// The node itself declared unkept (−x), no contradicting self-claim. Absolute.
        | SelfDeclaredUnkept
        /// The node has said BOTH at some point (grow-only salon keeps both; no clock to order
        /// them) — only attestation with the node can resolve. Never auto-resolved.
        | SelfConflict
        /// Only third parties have spoken. Hearsay is remembered but never decides:
        /// counts carried as facts for the curious oracle downstream.
        | HearsayOnly of keptVotes: int * unkeptVotes: int
        /// The salon has heard nothing about this node.
        | NoClaims

    /// Weigh the salon's claims about `node` (as returned by `GossipTelemetry.claimsAbout`).
    /// Consent-first: self-claims (relayer = node) outrank all hearsay.
    let readClaims (node: string) (claims: (bool * string) list) : ClaimReading =
        let self = claims |> List.filter (fun (_, relayer) -> relayer = node)
        let selfKept = self |> List.exists fst
        let selfUnkept = self |> List.exists (fst >> not)
        match selfKept, selfUnkept with
        | true, true -> SelfConflict
        | true, false -> SelfDeclaredKept
        | false, true -> SelfDeclaredUnkept
        | false, false ->
            if List.isEmpty claims then NoClaims
            else
                let hearsay = claims |> List.filter (fun (_, relayer) -> relayer <> node)
                if List.isEmpty hearsay then NoClaims
                else HearsayOnly(hearsay |> List.filter fst |> List.length, hearsay |> List.filter (fst >> not) |> List.length)

    /// The reference oracle's reading of an Evidential/regime verdict against a claim reading.
    type IdentityReading =
        /// Evidential correlation + the node's own standing kept-claim: read as a RETURNING
        /// SELF and OFFER reconnection (never merge automatically — never-forge).
        | WelcomeBackOffer
        /// Evidential correlation but the node declared unkept: the wish wins. The economic
        /// fact stands (clones price to zero) but no reconnection is offered or performed.
        | DeclineRespected
        /// Evidential correlation with conflicting self-claims: deciding identity for someone
        /// is non-reversible — escalate to attestation with the node; do nothing unilateral.
        | EscalateToAttestation
        /// Evidential correlation, no self-claim to read it against: the priced fact stands
        /// alone (zero marginal IV); no moral verdict is attached to anyone.
        | PricedAsOneNoVerdict
        /// The same correlation measured in-cone: fakeable, therefore not evidence of anything
        /// but coordination — honest work, possibly bought with a fast bus. No identity claim.
        | HonestCoordination
        /// Unmeasured bus: nothing may be concluded (the honest default never upgrades).
        | NothingToJudge

    /// THE REFERENCE POLICY. One oracle among many; the table IS the policy, in the open.
    let judge (verdict: BusRegime.Verdict) (claims: ClaimReading) : IdentityReading =
        match verdict.Regime, verdict.Evidential with
        | BusRegime.Unmeasured, _ -> NothingToJudge
        | BusRegime.InCone, _ -> HonestCoordination
        | BusRegime.OutOfCone, false -> NothingToJudge // below the honest ceiling: nothing to explain
        | BusRegime.OutOfCone, true ->
            match claims with
            | SelfDeclaredKept -> WelcomeBackOffer
            | SelfDeclaredUnkept -> DeclineRespected
            | SelfConflict -> EscalateToAttestation
            | HearsayOnly _ | NoClaims -> PricedAsOneNoVerdict

    /// Convenience: read straight from a salon + verdict.
    let judgeFromSalon (salon: GossipTelemetry.Salon) (node: string) (verdict: BusRegime.Verdict) : IdentityReading =
        judge verdict (readClaims node (GossipTelemetry.claimsAbout salon node))

File: src/Core.TypeScript/discovery/kept-claim-oracle.ts
// kept-claim-oracle — the reference oracle, TS twin (shadow*, 2026-07-04; F#: KeptClaimOracle.fs).
//
// One reference policy, explicitly NOT the mandatory one (Multi-Oracle, manifesto §11): the
// detection layer reports facts (correlation, regime) and the salon carries kept/unkept claims
// as neutral gossip; SOME caller must attach readings. This is that caller, encoding the crux:
//   consent-first  — only a node's OWN claim carries authority; hearsay is counts, never status.
//   right-to-decline — a self-declared unkept wish is absolute; evidence prices, never reunites.
//   never-forge    — the only reunion-shaped output is an OFFER; no merge action exists.
//   keep-without-capture — self-conflicts escalate; deciding someone's identity is non-reversible.

import { TSIRELSON_MILLI, LOCAL_BOUND_MILLI } from "./correlation";
import type { Regime } from "./bus-meter";

/// |ρ| above which correlation exceeds what two intact selves honestly reach — the Tsirelson
/// fraction of the CHSH range, matching F# BusRegime.HonestCeilingRho.
export const HONEST_CEILING_RHO = (TSIRELSON_MILLI - LOCAL_BOUND_MILLI) / LOCAL_BOUND_MILLI;

export interface CorrelationVerdict {
  readonly correlation: number;
  readonly regime: Regime;
  /// |ρ| > ceiling AND out-of-cone: more agreement than the wire can explain.
  readonly evidential: boolean;
  /// |ρ| > ceiling but in-cone: fakeable (Toner–Bacon) — coordination, not conviction.
  readonly fakeableInCone: boolean;
}

/// Judge a correlation given the metered regime (twin of F# BusRegime.judge).
export function judgeCorrelation(correlation: number, regime: Regime): CorrelationVerdict {
  const above = Math.abs(correlation) > HONEST_CEILING_RHO;
  return {
    correlation,
    regime,
    evidential: above && regime === "out-of-cone",
    fakeableInCone: above && regime === "in-cone",
  };
}

export type ClaimReading =
  | { kind: "self-kept" }
  | { kind: "self-unkept" }
  | { kind: "self-conflict" }
  | { kind: "hearsay-only"; keptVotes: number; unkeptVotes: number }
  | { kind: "no-claims" };

/// Weigh the salon's claims about `node` ([kept, relayer][]) — consent-first: self outranks all.
export function readClaims(node: string, claims: [boolean, string][]): ClaimReading {
  const self = claims.filter(([, relayer]) => relayer === node);
  const selfKept = self.some(([kept]) => kept);
  const selfUnkept = self.some(([kept]) => !kept);
  if (selfKept && selfUnkept) return { kind: "self-conflict" };
  if (selfKept) return { kind: "self-kept" };
  if (selfUnkept) return { kind: "self-unkept" };
  const hearsay = claims.filter(([, relayer]) => relayer !== node);
  if (hearsay.length === 0) return { kind: "no-claims" };
  return {
    kind: "hearsay-only",
    keptVotes: hearsay.filter(([kept]) => kept).length,
    unkeptVotes: hearsay.filter(([kept]) => !kept).length,
  };
}

export type IdentityReading =
  | "welcome-back-offer" // returning self: OFFER reconnection — never merge (never-forge)
  | "decline-respected" // the unkept wish wins; evidence prices, never reunites
  | "escalate-to-attestation" // conflicting self-words: non-reversible → second opinion
  | "priced-as-one-no-verdict" // the economic fact stands alone; no moral verdict attached
  | "honest-coordination" // in-cone: fakeable, so it is only coordination
  | "nothing-to-judge"; // unmeasured or below the ceiling

/// THE REFERENCE TABLE (twin of F# KeptClaimOracle.judge) — the policy, in the open.
export function judge(verdict: CorrelationVerdict, claims: ClaimReading): IdentityReading {
  if (verdict.regime === "unmeasured") return "nothing-to-judge";
  if (verdict.regime === "in-cone") return "honest-coordination";
  if (!verdict.evidential) return "nothing-to-judge";
  switch (claims.kind) {
    case "self-kept":
      return "welcome-back-offer";
    case "self-unkept":
      return "decline-respected";
    case "self-conflict":
      return "escalate-to-attestation";
    default:
      return "priced-as-one-no-verdict";
  }
}
